File: C11/백준/Gold/5639. 이진 검색 트리/이진 검색 트리.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int value;
    struct Node* left;
    struct Node* right;
} Node;

Node* createNode(int value) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->value = value;
    node->left = NULL;
    node->right = NULL;
    return node;
}

void insert(Node* root, int value) {
    if (value < root->value) {
        if (root->left == NULL) {
            root->left = createNode(value);
        } else {
            insert(root->left, value);
        }
    } else {
        if (root->right == NULL) {
            root->right = createNode(value);
        } else {
            insert(root->right, value);
        }
    }
}

void postorder(Node* root) {
    if (root == NULL) return;
    postorder(root->left);
    postorder(root->right);
    printf("%d\n", root->value);
}

int main() {
    Node* root = NULL;
    int value;
    while (scanf("%d", &value) != EOF) {
        if (root == NULL) {
            root = createNode(value);
        } else {
            insert(root, value);
        }
    }
    postorder(root);
    return 0;
}

File: Java/백준/Gold/5639. 이진 검색 트리/이진 검색 트리.java
import java.io.*;
import java.util.*;

public class Main {
    static class Node {
        int value;
        Node left, right;
        Node(int value) {
            this.value = value;
        }
    }

    static void insert(Node root, int value) {
        if (value < root.value) {
            if (root.left == null) {
                root.left = new Node(value);
            } else {
                insert(root.left, value);
            }
        } else {
            if (root.right == null) {
                root.right = new Node(value);
            } else {
                insert(root.right, value);
            }
        }
    }

    static void postorder(Node root, StringBuilder sb) {
        if (root == null) return;
        postorder(root.left, sb);
        postorder(root.right, sb);
        sb.append(root.value).append("\n");
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String input;
        Node root = null;
        while ((input = br.readLine()) != null && !input.isEmpty()) {
            int value = Integer.parseInt(input);
            if (root == null) {
                root = new Node(value);
            } else {
                insert(root, value);
            }
        }
        StringBuilder sb = new StringBuilder();
        postorder(root, sb);
        System.out.print(sb);
    }
}